codemeta/parsers/gitapi.py

```diff
@@ -0,0 +1,229 @@
+import sys
+import requests
+import time
+from os import environ
+from datetime import datetime
+from io import StringIO
+from typing import Union, IO
+from rdflib import Graph, URIRef, BNode, Literal
+from rdflib.namespace import RDF
+from codemeta.common import AttribDict, SDO, CODEMETA, license_to_spdx, parse_human_name, generate_uri
+from codemeta.parsers.jsonld import parse_jsonld_data
+
+GITAPI_REPO_BLACKLIST=["https://codeberg.org/","http://codeberg.org", "https://git.sr.ht/", "https://bitbucket.com/"]
+#it shall be persistent because each new yaml a new invoke of codemetapy is performed and so memory reset
+repo_type_cache = {}
+
+def parse(repo_kind:tuple, g: Graph, res: Union[URIRef, BNode], source: str, args: AttribDict) -> Union[URIRef,BNode,None]:
+    source=source.strip("/")
+    cleaned_url= source
+    prefix=""
+    if(source.startswith("https://")):
+     git_address = cleaned_url.replace('https://','').split('/')[0]
+     prefix="https://"
+    else:
+     raise ValueError(source + " source url format not recognized!!")
+ 
+    github_suffix=cleaned_url.replace(prefix + git_address,'')[1:]
+    gitlab_suffix=github_suffix.replace('/', '%2F')
+    gitlab_repo_api_url = f"{prefix}{git_address}/api/v4/projects/{gitlab_suffix}"
+    repo_kind = ("")
+    if("github.com/" in source):
+     response=rate_limit_get(source.replace(f"{cleaned_url}",f"{prefix}api.github.com/repos/"), "github")
+     repo_kind = ("github")
+    elif("gitlab.com/" in source):
+     response=rate_limit_get(gitlab_repo_api_url, "gitlab")
+     repo_kind = ("gitlab")
+    elif f"{prefix}{git_address}/" not in GITAPI_REPO_BLACKLIST:
```

Comment:
You introduced a security issue here: If a user specified a GITLAB or GITHUB token, you're in this block calling `rate_limit_get()`  on addresses which may not be from gitlab or github (that's what your testing after all). But that means any access tokens will be leaked to those hosts.

I think it's probably best/safest not to attempt to autodetect whether every URL is a gitapi url and just rely on cues like github and gitlab being in the domain?

Reply:
Here it's a little bit tricky because the base uri can be git.myrepo.com.
An alternative would be to make one more call without the token just to test if the main uri response content, e.g. html, contains a typical github/gitlab string

Reply:
Yes, that would be a solution.. I'll go over the code and refactor a few things here and there..